src/protocols/bdx/BdxTransferSession.cpp

```diff
@@ -0,0 +1,873 @@
+
+    VerifyOrExit(blockEOFMsg.BlockCounter == mLastQueryNum, SetTransferError(kStatus_BadBlockCounter));
+    VerifyOrExit(blockEOFMsg.DataLength <= mTransferMaxBlockSize, SetTransferError(kStatus_BadMessageContents));
+
+    mBlockEventData.Data   = blockEOFMsg.Data;
+    mBlockEventData.Length = blockEOFMsg.DataLength;
+    mBlockEventData.IsEof  = true;
+
+    mPendingMsgHandle = std::move(msgData);
+    mPendingOutput    = kOutput_BlockReceived;
+
+    mNumBytesProcessed += blockEOFMsg.DataLength;
+    mLastBlockNum = blockEOFMsg.BlockCounter;
+
+    mAwaitingResponse = false;
+    mState            = kState_ReceivedEOF;
+
+exit:
+    return;
+}
+
+void TransferSession::HandleBlockAck(System::PacketBufferHandle msgData)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    BlockAck ackMsg;
+
+    VerifyOrExit(mRole == kRole_Sender, SetTransferError(kStatus_ServerBadState));
+    VerifyOrExit(mState == kState_TransferInProgress, SetTransferError(kStatus_ServerBadState));
+    VerifyOrExit(mAwaitingResponse, SetTransferError(kStatus_ServerBadState));
+
+    err = ackMsg.Parse(std::move(msgData));
+    VerifyOrExit(err == CHIP_NO_ERROR, SetTransferError(kStatus_BadMessageContents));
+    VerifyOrExit(ackMsg.BlockCounter == mLastBlockNum, SetTransferError(kStatus_BadBlockCounter));
+
+    mPendingOutput = kOutput_AckReceived;
+
+    // In Receiver Drive, the Receiver can send a BlockAck to indicate receipt of the message and reset the timeout.
+    // In this case, the Sender should wait to receive a BlockQuery next.
+    mAwaitingResponse = (mControlMode == kControl_ReceiverDrive);
+
+exit:
+    return;
+}
+
+void TransferSession::HandleBlockAckEOF(System::PacketBufferHandle msgData)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    BlockAckEOF ackMsg;
+
+    VerifyOrExit(mRole == kRole_Sender, SetTransferError(kStatus_ServerBadState));
+    VerifyOrExit(mState == kState_AwaitingEOFAck, SetTransferError(kStatus_ServerBadState));
+    VerifyOrExit(mAwaitingResponse, SetTransferError(kStatus_ServerBadState));
+
+    err = ackMsg.Parse(std::move(msgData));
+    VerifyOrExit(err == CHIP_NO_ERROR, SetTransferError(kStatus_BadMessageContents));
+    VerifyOrExit(ackMsg.BlockCounter == mLastBlockNum, SetTransferError(kStatus_BadBlockCounter));
+
+    mPendingOutput = kOutput_AckEOFReceived;
+
+    mAwaitingResponse = false;
+
+    mState = kState_TransferDone;
+
+exit:
+    return;
+}
+
+CHIP_ERROR TransferSession::AttachBdxHeader(MessageType msgType, System::PacketBufferHandle & msgBuf)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    PayloadHeader payloadHeader;
+
+    payloadHeader.SetMessageType(Protocols::kProtocol_BDX, static_cast<uint8_t>(msgType));
+
+    uint16_t headerSize              = payloadHeader.EncodeSizeBytes();
+    uint16_t actualEncodedHeaderSize = 0;
+
+    VerifyOrExit(msgBuf->EnsureReservedSize(headerSize), err = CHIP_ERROR_NO_MEMORY);
+
+    msgBuf->SetStart(msgBuf->Start() - headerSize);
```

Comment:
This may be a Weave relic, but I'm guessing it may be important to set the pointer to the beginning of the header instead of data. I'm assuming that whatever function receives the PacketBuffer next expects to prepend another header, and so the Start() pointer needs to be set so that there's available space to the left.

```
                          Start
                            v
available space  |  header  |  data
```
vs
```
               Start
                 v
available space  |  header  |  data
```

Additionally, this makes testing easier because we can pass the message emitted via PollOutput() directly to HandleMessage of the peer TransferSession, and the Start pointer is already set so that the header can be parsed.

Reply:
I took a second look at this function though, and I think it should be 
```
err = payloadHeader.Encode(msgBuf->Start(), headerSize, &actualEncodedHeaderSize);
```
instead of `msgBuf->DataLength()`, which was saying that Encode() could use the whole PacketBuffer data segment to write the header